include/dlaf/permutations/general.h

```diff
@@ -36,25 +36,28 @@ namespace dlaf::permutations {
 template <Backend B, Device D, class T, Coord coord>
 void permute(SizeType i_begin, SizeType i_end, Matrix<const SizeType, D>& perms,
              Matrix<const T, D>& mat_in, Matrix<T, D>& mat_out) {
-  const matrix::Distribution& distr_perms = perms.distribution();
-  const matrix::Distribution& distr_in = mat_in.distribution();
-  const matrix::Distribution& distr_out = mat_out.distribution();
-
   DLAF_ASSERT(matrix::local_matrix(perms), perms);
   DLAF_ASSERT(matrix::local_matrix(mat_in), mat_in);
   DLAF_ASSERT(matrix::local_matrix(mat_out), mat_out);
 
+  // Note:
+  // These are not implementation constraints, but more logic constraints. Indeed, these ensure that
+  // the range [i_begin, i_end] is square in terms of elements (it would not make sense to have it square
+  // in terms of number of tiles). Moreover, by requiring mat_in and mat_out matrices to have the same
+  // shape, it is ensured that range [i_begin, i_end] is actually the same on both sides.
+  DLAF_ASSERT(square_size(mat_in), mat_in);
+  DLAF_ASSERT(matrix::square_blocksize(mat_in), mat_in);
+  DLAF_ASSERT(matrix::equal_size(mat_in, mat_out), mat_in);
+  DLAF_ASSERT(matrix::equal_blocksize(mat_in, mat_out), mat_in);
+
   DLAF_ASSERT(i_begin >= 0 && i_begin <= i_end, i_begin, i_end);
 
-  DLAF_ASSERT(i_end <= distr_perms.nrTiles().rows(), i_end, perms);
-  DLAF_ASSERT(i_end <= distr_in.nrTiles().rows() && i_end <= distr_in.nrTiles().cols(), i_end, mat_in);
-  DLAF_ASSERT(i_end <= distr_out.nrTiles().rows() && i_end <= distr_out.nrTiles().cols(), i_end,
-              mat_out);
+  DLAF_ASSERT(i_end <= perms.nrTiles().rows(), i_end, perms);
```

Comment:
A similar problem to what can happen with different matrix sizes might happens if the permutation has different size. I would add a check that the number of rows of the permutation has to match the size of the matrix.